sklearn/neighbors/lmnn.py

```diff
@@ -0,0 +1,1070 @@
+        self.verbose = verbose
+        self.random_state = random_state
+        self.n_jobs = n_jobs
+
+    def fit(self, X, y):
+        """Fit the model according to the given training data.
+
+        Parameters
+        ----------
+        X : array-like, shape (n_samples, n_features)
+            The training samples.
+
+        y : array-like, shape (n_samples,)
+            The corresponding training labels.
+
+        Returns
+        -------
+        self : object
+            returns a trained LargeMarginNearestNeighbor model.
+        """
+
+        # Validate the inputs
+        X, y = check_X_y(X, y, ensure_min_samples=2)
+        check_classification_targets(y)
+
+        # Check that the inputs are consistent with the parameters
+        X_valid, y_valid, init = self._validate_params(X, y)
+
+        # Initialize the random generator
+        self.random_state_ = check_random_state(self.random_state)
+
+        # Measure the total training time
+        t_train = time.time()
+
+        # Initialize the linear transformation
+        transformation = self._initialize(X_valid, init)
+
+        # Find the target neighbors
+        targets = _select_target_neighbors(X_valid, y_valid, self.n_neighbors_,
+                                           verbose=self.verbose,
+                                           n_jobs=self.n_jobs,
+                                           algorithm=self.neighbors_algorithm)
+
+        # Compute the gradient part contributed by the target neighbors
+        grad_static = _compute_grad_static(X_valid, targets, self.verbose)
+
+        # Decide how to store the impostors
+        if self.imp_store == 'sparse':
+            use_sparse = True
+        elif self.imp_store == 'list':
+            use_sparse = False
+        else:
+            # auto: Use a heuristic based on the data set size
+            use_sparse = X_valid.shape[0] > 6500
+
+        # Create a dictionary of parameters to be passed to the optimizer
+        disp = self.verbose - 2 if self.verbose > 1 else -1
+        optimizer_params = {'method': 'L-BFGS-B',
+                            'fun': self._loss_grad_lbfgs,
+                            'jac': True,
+                            'args': (X_valid, y_valid, targets, grad_static,
+                                     use_sparse),
+                            'x0': transformation,
+                            'tol': self.tol,
+                            'options': dict(maxiter=self.max_iter, disp=disp),
+                            'callback': self._callback
+                            }
+
+        # Call the optimizer
+        self.n_iter_ = 0
+        opt_result = minimize(**optimizer_params)
+
+        # Reshape the solution found by the optimizer
+        self.transformation_ = opt_result.x.reshape(-1, X_valid.shape[1])
+
+        # Stop timer
+        t_train = time.time() - t_train
+        if self.verbose:
+            print('Training took {:8.2f}s.'.format(t_train))
```

Comment:
Here you should reference the class name of where this message originates from: think in a large codebase with many things going on. You can retrieve the class name with self.__class__.__name__.

Reply:
Done.